Prototype ghost cells for use with foyle

jlewi opened the pull request:
This is an initial implementation of having the AI autogenerate new cells as a user types. This means the user doesn't have to explicitly ask for the AI to generate cells.

This is currently disabled by default but can be enabled with an experimental flag: "AI AutoComplete".

The feature seems to be working with a couple rough spots

* Per jlewi/foyle#168 cell decorations can't be applied to markdown cells; so markdown cells just look like regular cells not ghost cells which is confusing
* AIQuality is an issue jlewi/foyle#170 we will really want to tune the AI to generate more relevant completions



Refer to [TN008 Automatically Insert Cells](https://github.com/jlewi/foyle/blob/main/tech_notes/tn008_auto_insert_cells.md)

Related to: jlewi/foyle#167

Here's a demo
https://github.com/user-attachments/assets/2d68fbd3-8d3b-43d5-accc-5f17c723dea2

## How streaming works

The streaming protocol defines two types of messages

1. Full context
2. Incremental update

The full context is only sent on the initial request. Subsequent messages only need to include the delta.

In the initial implementation, the full context includes the full notebook. This could end up being quite large O(Mb) for large notebooks with verbose outputs. We may want to reduce this in the future. 

Incremental updates are just the latest version of the current cell. We resend the latest version of the current cell because its more robust and easier than trying to send just a delta to the current cell.

The expectation is that a streaming request is initiatied when a user starts editing a cell causing the server to generate a new series of ghost cells based on their input. When the focus switches to a new cell that stream would get closed and a new one would be initiated.

However, the connection can be closed prematurely for a variety of reasons. The biggest one being server timeouts. In this case we need to restart the connection to continue generating completions based on the cell currently being edited.


## Why not use observable?

I played around with observable and there were two reasons why I eventually got rid of it.

The connect protocol uses AsyncIterable which are pull semantics whereas Observable is push semantics. I found converting between them to be difficult and confusing.

I think the bigger problem was that I was originally hoping to use Observable to so stream processing. I wanted to treat VSCode events as a stream and do things like windowing and rate limiting. My initial idea was to use windowing to divide the stream of events into distinct rpc streams. I found the windowing semantics very confusing and I'm not sure they were doing what I wanted.

The problem I ran into with the streaming model is handling connection failures. If there is a connection error (e.g. closed due to timeout) this results in an exception when trying to read the next response. We handle this by starting a new stream. I couldn't figure out how to model this in observable because this seems to violate the "fire and forget" nature of stream processing.

jlewi commented:
The current PR demonstrates feasibility of using GhostCells.
It shouldn't be merged as is because its not guarded by a flag. We should figure out how to guard it with a flag and then we could potentially merge it. Then in subsequent iterations we could wire it up to the Foyle server.

jlewi commented:
So `callStreamGenerate` is not working.

Here's my problem. I'm trying to implement streaming code completion. So I have a stream of strings
const data = [
'hello',
'how are you?',
'<stop>',
"Is it me you're looking for?",
'<stop>',
'here we go',
'again',
'down the only road I have ever known',
]
The <stop> sequence indicates a segmentation boundary at which I want to divide the stream into sub streams e.g.
[    ["hello", "how are you?", '<stop>'],    ["Is it me you're looking for?", '<stop>']    ["here we go", "again", "down the only road I have ever known"],]
Each sub array e.g. ["hello", "how are you", 'stop'] corresponds to one streaming RPC call. So for each element in this array e.g. "hello" I want to stream a request message. I will also get back a stream of responses.
I'm modeling the data using an Observable<string> because vscode events have a push model.
However, my streaming RPC client using AsyncIterable. I'm having trouble figuring out how to glues these two together.
My latest attempt is [here](https://github.com/stateful/vscode-runme/blob/b3cf19396e8ef1c90db89fc46e77077609adf015/src/extension/ai/stream.ts#L208). I'm using Observable's windowing functionality to turn Observable<string> into Observable<Observable<string>> . I want to turn eachObservable<string> into an AsyncIterable<StreamRequest>.

The current implementation isn't doing what we want.

There's a couple problems.

1. We need to introduce a delay between the data items. This simulates a delay as a user types. If we don't then all of the items pile up waiting for the responses. So we end up issuing multiple requests in parallel which isn't what would normally happen.

2. Right now the responses don't actually get printed. I don't think my handling of the responses is correct. The AsyncIterable<StreamResponse> is a pull model so I'm not sure how to deal with that.

jlewi commented:
I think this is ready for review.

@pastuxso I was wondering if you could help me resolve the build error
https://gist.github.com/jlewi/f3e4fb85b0b5d716f772aa898ea25fe7
That is causing the tests to fail.

The extension works when I debug it but `npm run build` fails with the given error

pastuxso commented:
@jlewi Do you still need my help on the build error? It worked fine for me.

jlewi commented:
@pastuxso Thank you I think I figured out the bundle error.

jlewi commented:
@sourishkrout I think this is ready for review.

There's a test failure.

```

 FAIL  tests/extension/serializer.test.ts [ tests/extension/serializer.test.ts ]

TypeError: window.createOutputChannel is not a function

 ❯ src/extension/logger.ts:7:30

      5| import { stripANSI } from '../utils/ansi'

      6| 

      7| const outputChannel = window.createOutputChannel('Runme')

       |                              ^

      8| 

      9| const DEFAULT_LOG_LEVEL: LogLevel = 'info'

 ❯ src/extension/ai/ghost.ts:3:31

```

I think the problem is here
https://github.com/stateful/vscode-runme/blob/454601d9675cd01892119f04ec9033bb8383cd46/src/extension/logger.ts#L7

IUC having toplevel definitions of variables that call vscode functions interferes with the vi mocking library.

It looks like my code triggers this because I'm importing the loggger function.

Do you have suggestions on how to to fix it?

I think one option would be to set it to null and then initialize it in getLogger if its not already initialized. WDYT?

sourishkrout commented:
> IUC having toplevel definitions of variables that call vscode functions interferes with the vi mocking library.
> 
> It looks like my code triggers this because I'm importing the loggger function.
> 
> Do you have suggestions on how to to fix it?
> 
> I think one option would be to set it to null and then initialize it in getLogger if its not already initialized. WDYT?

I'll take a look and let you know as part of the review. I believe there's likely a way to solve this with the mocking library that is non-obvious but I have been in similar scenarios before.

sourishkrout commented:
> > IUC having toplevel definitions of variables that call vscode functions interferes with the vi mocking library.
> > It looks like my code triggers this because I'm importing the loggger function.
> > Do you have suggestions on how to to fix it?
> > I think one option would be to set it to null and then initialize it in getLogger if its not already initialized. WDYT?
> 
> I'll take a look and let you know as part of the review. I believe there's likely a way to solve this with the mocking library that is non-obvious but I have been in similar scenarios before.

@jlewi I'm fairly certain inserting this one line will make the tests pass:

```
diff --git a/tests/extension/serializer.test.ts b/tests/extension/serializer.test.ts
index 1a17823a..57c2bd77 100644
--- a/tests/extension/serializer.test.ts
+++ b/tests/extension/serializer.test.ts
@@ -28,6 +28,7 @@ vi.mock('vscode', () => ({
   window: {
     activeNotebookEditor: undefined,
     showErrorMessage: vi.fn().mockResolvedValue({}),
+    createOutputChannel: vi.fn(),
   },
   Uri: { joinPath: vi.fn().mockReturnValue('/foo/bar') },
   workspace: {
```

sourishkrout commented on src/extension/serializer.ts:
```diff
@@ -186,7 +187,17 @@ export abstract class SerializerBase implements NotebookSerializer, Disposable {
-    data = new NotebookData(cells)
+
+    // Prune any ghost cells when saving.
+    let cellsToSave = []
```
nit, this should be const. what's constant is the handle (unassigable) to the array whereas it's elements are mutable.

sourishkrout replied:
drastic modification of the markdown document at serialization time will likely cause a reload in vscode which wipes editor state. wondering if `onWillSave...` allows to use the WorkspaceEdit APIs to make this transparent to the user/UX.

sourishkrout replied:
this looks promising; we might have to bump to a higher minimum VS Code version (which is fine) to get this API:

https://github.com/microsoft/vscode/blob/1.92.1/src/vscode-dts/vscode.d.ts#L13519

jlewi replied:
> drastic modification of the markdown document at serialization time will likely cause a reload in vscode 

I'm not sure I fully follow you about the problem and the proposed solution.

My thinking was that we want the in memory/rendered version to be different from the on disk version. So ghost cells should appear in the memory version but not the on disk representation.

I thought the way vscode serialization worked was VSCode invoked the serializer with the data NoteBookCellData  structure representing the data and then the serializer gets back a stream of bytes which vscode then writes to the file.  So I was thinking that as part of the serializer we could filter out the ghost cells.  It sounds like your saying that VSCode can detect if the serializer modifies the byte stream and if it does trigger a reload?

Maybe I just implemented it at the wrong place? For example, if we implemented the filtering in the golang serializer how would vscode know that the byte stream it was getting back.

How would using onWillSave and WorkspaceEdit solve this problem? If we use onWillSave and WorkspaceEdit to remove the ghost cells I think the removal of the deletion of the ghost cells would be visible to the users? What happens in the "auto save" case. Consider the following use case.

1) I start editing a markdown cell
2) Ghost cells are rendered
3) I save the file  or file gets auto saved

I think the desired behavior is
i) serialized notebook doesn't contain the ghost cells
ii) ghost cells continue to show up in the UI

sourishkrout replied:
gotcha. i think i may have misunderstood and taken the "save" to literal. In my mind "save" is a user action whereas serialization is scheduled by vscode and not exclusively triggered by saving a dirty document. thanks for the clarification, i know understand the purpose of the code is to skip serializing ghost cells unless they were previously accepted.

sourishkrout commented on src/extension/serializer.ts:
```diff
@@ -870,3 +881,8 @@ export class GrpcSerializer extends SerializerBase {
   }
 }
+
+function isGhostCell(cell: NotebookCellData): boolean {
```
must utility methods on the Serializer are static methods. use same pattern for consistency?

jlewi replied:
Done.

sourishkrout reviewed (commented):
Pretty close to something merge-able. Left some comments.

jlewi commented on package.json:
```diff
@@ -1134,7 +1141,10 @@
+    "@buf/jlewi_foyle.bufbuild_es": "^1.10.0-20240801044500-fba3b5ceb26b.1",
+    "@buf/jlewi_foyle.connectrpc_es": "^1.4.0-20240801044500-fba3b5ceb26b.3",
     "@buf/stateful_runme.community_timostamm-protobuf-ts": "^2.9.4-20240731204114-2df61af8e022.4",
+    "@bufbuild/connect-node": "^0.13.0",
```
Done.

sourishkrout replied:
fixed the imports in a separate PR

jlewi commented on src/extension/ai/ghost.ts:
```diff
@@ -0,0 +1,387 @@
+
+  // Create a stream creator. The StreamCreator is a class that effectively windows events
+  // and turns each window into an AsyncIterable of streaming requests.
+  let creator = new stream.StreamCreator(cellGenerator)
```
Thanks. I updated the PR description to document some of the reasons why I didn't use Observable.

sourishkrout replied:
Makes sense. You have something that works 👍 so let's roll with it.

jlewi commented on src/extension/ai/stream.ts:
```diff
@@ -0,0 +1,267 @@
+import getLogger from '../logger'
+
+const log = getLogger()
+const baseUrl = 'http://localhost:8080/api'
```
Done.

jlewi commented on src/extension/extension.ts:
```diff
@@ -162,6 +162,12 @@ export class RunmeExtension {
+    const aiManager = new manager.AIManager(context)
+    // We need to hang onto a reference to the AIManager so it doesn't get garbage collected until the
+    // extension is deactivated.
+    context.subscriptions.push(aiManager)
```
Should be fixed.

jlewi commented on src/extension/ai/manager.ts:
```diff
@@ -0,0 +1,17 @@
+import * as ghost from './ghost'
+
+// AIManager is a class that manages the AI services.
+export class AIManager {
```
I refactored the code please take a look and let me know if its what you meant.

sourishkrout replied:
You got it

jlewi commented on src/extension/ai/converters.ts:
```diff
@@ -23,3 +28,34 @@ export function cellToCellData(cell: vscode.NotebookCell): vscode.NotebookCellDa
 }
+
+// cellProtosToCellData converts an array of RunMe cell protos to an array of VSCode CellData
+export function cellProtosTocellData(cells: serializerTypes.Cell[]): vscode.NotebookCellData[] {
```
Done.

jlewi commented:
@sourishkrout PTAL. I think I addressed most of your comments. There's one substantial issue pending which is how should we handle filtering out the ghost cells on save.

sourishkrout commented:
> @sourishkrout PTAL. I think I addressed most of your comments. There's one substantial issue pending which is how should we handle filtering out the ghost cells on save.

👍 thank you. Granted https://github.com/stateful/vscode-runme/pull/1535 passes and is merged into this PR, we should be good to merge.

sourishkrout commented:
> ## Why not use observable?
> I played around with observable and there were two reasons why I eventually got rid of it.
> 
> The connect protocol uses AsyncIterable which are pull semantics whereas Observable is push semantics. I found converting between them to be difficult and confusing.
> 
> I think the bigger problem was that I was originally hoping to use Observable to so stream processing. I wanted to treat VSCode events as a stream and do things like windowing and rate limiting. My initial idea was to use windowing to divide the stream of events into distinct rpc streams. I found the windowing semantics very confusing and I'm not sure they were doing what I wanted.
> 
> The problem I ran into with the streaming model is handling connection failures. If there is a connection error (e.g. closed due to timeout) this results in an exception when trying to read the next response. We handle this by starting a new stream. I couldn't figure out how to model this in observable because this seems to violate the "fire and forget" nature of stream processing.

Totally fair 👍. For clarification, rxjs's Subject, plus the use of higher-order observables in combination with transformation operators, can map those semantics concisely. However, this is observables++ territory with a significant on-ramp, and now that we have a working implementation, let's not dwell on it.

sourishkrout commented:
> > @sourishkrout PTAL. I think I addressed most of your comments. There's one substantial issue pending which is how should we handle filtering out the ghost cells on save.
> 
> 👍 thank you. Granted #1535 passes and is merged into this PR, we should be good to merge.

@jlewi #1535 passed, and we can merge it into this PR. The clarification for "filtering out the ghost cells on save" makes sense. It should be good to merge once the tests pass.
